qcommon/src/main/java/jake2/qcommon/network/Netchan.java

```diff
@@ -177,68 +156,79 @@ public static boolean Netchan_NeedReliable(netchan_t chan) {
-                    + ":Outgoing message overflow\n");
+            Com.Printf(chan.remote_address.toString() + ":Outgoing message overflow\n");
             return;
         }
 
-        int send_reliable = Netchan_NeedReliable(chan) ? 1 : 0;
+        int send_reliable = chan.needReliable() ? 1 : 0;
 
-        if (chan.reliable_length == 0 && chan.message.cursize != 0) {
-            System.arraycopy(chan.message_buf, 0, chan.reliable_buf, 0,
-                    chan.message.cursize);
-            chan.reliable_length = chan.message.cursize;
-            chan.message.cursize = 0;
+        if (chan.reliable_length == 0 && chan.reliable.size() != 0) {
+            // wrap chan.reliable_buf array in a buffer
+            final sizebuf_t reliableDataBuffer = new sizebuf_t();
+            reliableDataBuffer.data = chan.reliable_buf;
+            reliableDataBuffer.maxsize = Defines.MAX_MSGLEN - 16;
+            reliableDataBuffer.allowoverflow = true;
+
+            chan.reliable.forEach(networkMessage -> networkMessage.writeTo(reliableDataBuffer));
+            chan.reliable_length = pendingReliableSize;
+            chan.reliable.clear();
             chan.reliable_sequence ^= 1;
         }
 
-        // write the packet header
-        buffer.init(send_buf, send_buf.length);
+        sizebuf_t packet = new sizebuf_t();
+        byte[] send_buf = new byte[Defines.MAX_MSGLEN];
+        packet.init(send_buf, send_buf.length);
 
+        // write the packet header
         int sequence = (chan.outgoing_sequence & ~(1 << 31)) | (send_reliable << 31);
         int sequenceAck = (chan.incoming_sequence & ~(1 << 31)) | (chan.incoming_reliable_sequence << 31);
 
         chan.outgoing_sequence++;
         chan.last_sent = (int) Globals.curtime;
 
-        buffer.writeInt(sequence);
-        buffer.writeInt(sequenceAck);
+        packet.writeInt(sequence);
+        packet.writeInt(sequenceAck);
 
         // send the qport if we are a client
         if (chan.sock == Defines.NS_CLIENT)
-            buffer.writeShort((int) qport.value);
+            packet.writeShort((int) qport.value);
 
         // copy the reliable message to the packet first
         if (send_reliable != 0) {
-            buffer.writeBytes(chan.reliable_buf, chan.reliable_length);
+            packet.writeBytes(chan.reliable_buf, chan.reliable_length);
             chan.last_reliable_sequence = chan.outgoing_sequence;
         }
 
-        // add the unreliable part if space is available
-        if (buffer.maxsize - buffer.cursize >= length)
-            buffer.writeBytes(data, length);
-        else
-            Com.Printf("Netchan_Transmit: dumped unreliable\n");
+        if (unreliable != null) {
+            // add the unreliable part if space is available
+            int length = unreliable.stream().mapToInt(NetworkMessage::getSize).sum();
+            if (packet.maxsize - packet.cursize >= length) {
+                unreliable.forEach(msg -> msg.writeTo(packet));
```

Comment:
now clear as day

Reply:
Fun fact:  we discard all pending unreliable messages, if they don't fit into the packet together, don't try to send some of them